### PR TITLE
NuScenes dataparser

### PR DESCRIPTION
**Pull request by zmurez**
This adds support for the NuScenes dataset #921. Optionally you can use the 3D cuboid labels to mask dynamic objects.

### CODE REVIEW

File: nerfstudio/data/dataparsers/nuscenes_dataparser.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+from dataclasses import dataclass, field
+import math
+import numpy as np
+import os
+from typing import Optional, Type, Tuple
+from typing_extensions import Literal
+from pathlib import Path
+import torch
+
+from nuscenes.nuscenes import NuScenes as NuScenesDatabase
```
Comment (tancik) [inline nerfstudio/data/dataparsers/nuscenes_dataparser.py:10]: This package is missing from the requirements.

File: scripts/process_nuscenes_masks.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+from dataclasses import dataclass
```
Comment (tancik) [inline scripts/process_nuscenes_masks.py:1]: Can you create a new folder `scripts/datasets/` to put this file in.

File: nerfstudio/data/dataparsers/nuscenes_dataparser.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+def rotation_translation_to_pose(r_vec, t_vec):
+    pose = np.eye(4)
+    pose[:3, :3] = qvec2rotmat(r_vec)
+    pose[:3, 3] = t_vec
+    return pose
+
+
+@dataclass
+class NuScenesDataParserConfig(DataParserConfig):
+    """Nerfstudio dataset config"""
```
Comment (tancik) [inline nerfstudio/data/dataparsers/nuscenes_dataparser.py:31]: Can you add a little more information about what the nuscenes dataset it, including a mention of the mask script. I worry people won't know the mask script exists.

**Review by tancik — COMMENTED** [review]
This is looking great! Can you add a screenshot to the PR showing a rendering of this dataset.

**Comment by zmurez** [comment]
```
ns-train nerfacto --vis tensorboard nuscenes-data --data-dir data/NuScenes --mask-dir data/NuScenes --data scene-0061
```

image
image

**Comment by tancik** [comment]
Can you run `ns-dev-test`, it will fix the build errors.

**Comment by zmurez** [comment]
> Can you run `ns-dev-test`, it will fix the build errors.

done

**Comment by segments-tobias** [comment]
Just saw the comment here:
`"waiting on multiple camera support" # TODO: remove once multiple cameras are supported`
so that answers my question ;)

**Comment by WUMINGCHAzero** [comment]
@zmurez Hi! Thanks for your great work! I wonder about the difference between world coordinates in nuScenes and that in nerfstudio. (I've tried to find it by myself but failed, still confused after reading the docs.) Thanks a lot!

File: nerfstudio/data/dataparsers/nuscenes_dataparser.py (hunk 3 above)
Comment (rockywind) [reply nerfstudio/data/dataparsers/nuscenes_dataparser.py]: Hi,
How to generate the mask of the Nuscene dataset.
````
[23:33:03] Saving config to: outputs/scene-0061/nerfacto/2023-02-01_233303/config.yml           experiment_config.py:124
[23:33:03] Saving checkpoints to: outputs/scene-0061/nerfacto/2023-02-01_233303/nerfstudio_models         trainer.py:120
logging events to: outputs/scene-0061/nerfacto/2023-02-01_233303
Setting up training dataset...
Caching all 36 images.
Printing profiling stats, from longest to shortest duration in seconds
Traceback (most recent call last):
  File "/home/zongmu/anaconda3/envs/nerfstudio/bin/ns-train", line 8, in <module>
    sys.exit(entrypoint())
  File "/DATA2/xiaoquan.wang/nerfstudio/scripts/train.py", line 247, in entrypoint
    main(
  File "/DATA2/xiaoquan.wang/nerfstudio/scripts/train.py", line 233, in main
    launch(
  File "/DATA2/xiaoquan.wang/nerfstudio/scripts/train.py", line 172, in launch
    main_func(local_rank=0, world_size=world_size, config=config)
  File "/DATA2/xiaoquan.wang/nerfstudio/scripts/train.py", line 86, in train_loop
    trainer.setup()
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/engine/trainer.py", line 145, in setup
    self.pipeline = self.config.pipeline.setup(
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/configs/base_config.py", line 57, in setup
    return self._target(self, **kwargs)
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/pipelines/base_pipeline.py", line 222, in __init__
    self.datamanager: VanillaDataManager = config.datamanager.setup(
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/configs/base_config.py", line 57, in setup
    return self._target(self, **kwargs)
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/datamanagers/base_datamanager.py", line 327, in __init__
    super().__init__()
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/datamanagers/base_datamanager.py", line 150, in __init__
    self.setup_train()
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/datamanagers/base_datamanager.py", line 360, in setup_train
    self.train_image_dataloader = CacheDataloader(
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/utils/dataloaders.py", line 80, in __init__
    self.cached_collated_batch = self._get_collated_batch()
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/utils/dataloaders.py", line 117, in _get_collated_batch
    batch_list = self._get_batch_list()
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/utils/dataloaders.py", line 111, in _get_batch_list
    batch_list.append(res.result())
  File "/home/zongmu/anaconda3/envs/nerfstudio/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/home/zongmu/anaconda3/envs/nerfstudio/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/zongmu/anaconda3/envs/nerfstudio/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/datasets/base_dataset.py", line 120, in __getitem__
    data = self.get_data(image_idx)
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/datasets/base_dataset.py", line 101, in get_data
    data["mask"] = get_image_mask_tensor_from_path(filepath=mask_filepath, scale_factor=self.scale_factor)
  File "/DATA2/xiaoquan.wang/nerfstudio/nerfstudio/data/utils/data_utils.py", line 29, in get_image_mask_tensor_from_path
    pil_mask = Image.open(filepath)
  File "/home/zongmu/anaconda3/envs/nerfstudio/lib/python3.8/site-packages/PIL/Image.py", line 3227, in open
    fp = builtins.open(filename, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/DATA/xinchao.gou/nuscenes/masks/CAM_FRONT/n015-2018-07-24-11-22-45+0800__CAM_FRONT__1532402930612460.png
````

**Comment by hjxwhy** [comment]
@zmurez Thanks for your contribution! Have you tested nuScenes dataset's other cameras? I have tested it with other cameras and all of them failed otherwise front camera. Thanks!

**Comment by donjiaking** [comment]
Hi, I wonder when to add multi-camera support on Nuscenes Dataset?

**Comment by donjiaking** [comment]
> @zmurez Thanks for your contribution! Have you tested nuScenes dataset's other cameras? I have tested it with other cameras and all of them failed otherwise front camera. Thanks!

@hjxwhy Hi, do you figure out how to use other cameras?

**Comment by mzlchou** [comment]
Is there any way to use multiple cameras yet?

**Pull request merged**